### ISSUE TITLE
Lottie animation not showing on iOS

### ISSUE THREAD

**Issue by santinucera** [lottie-react-native/lottie-react-native#581]
### Description

Hello, I'm having an issue with an asset, that isn't showing on iOS. I've recently updated `react-native` to `0.61.2`. Previously to this update this asset was showing up. In Android the asset is showing up as is intended. This only happen with this particularly [asset](https://assets8.lottiefiles.com/packages/lf20_FiUiqO.json). 

### Steps to Reproduce

```js

import React, { Component } from 'react';
import { Animated, Easing } from 'react-native';
import PropTypes from 'prop-types';
import LottieView from 'lottie-react-native';
import { sourceModel } from '@propTypes/sourceModel';

class SingleTimeAnimation extends Component {
  state = {
    progress: new Animated.Value(0)
  };

  componentDidMount() {
    this.animateLottie();
  }

  componentDidUpdate(prevProps) {
    const { source } = this.props;
    if (prevProps.source !== source) {
      this.resetAnimation();
    }
  }

  resetAnimation = () => {
    this.setState({ progress: new Animated.Value(0) }, this.animateLottie);
  };

  animateLottie = () => {
    const { progress } = this.state;
    const { duration } = this.props;
    Animated.timing(progress, {
      toValue: 1,
      duration,
      easing: Easing.linear,
      useNativeDriver: true
    }).start();
  };

  render() {
    const { progress } = this.state;
    const { source, style } = this.props;
    return <LottieView style={style} source={source} progress={progress} />;
  }
}

SingleTimeAnimation.propTypes = {
  duration: PropTypes.number.isRequired,
  source: sourceModel
};

export default SingleTimeAnimation;

```

### Versions

"react-native": "0.61.2"
"lottie-react-native": "^3.1.1"

**Comment by elgambet** [comment]
Hi, also happening in:

"react-native": "0.61.2"
"lottie-ios": "3.1.3",
"lottie-react-native": "3.2.1",

In my case the asset I'm using works in Android and worked in IOS with previous versions of react-native and lottie. 

Other animations are working without problems.

**Comment by JakeHadley** [comment]
I'm having this issue as well, I'm not sure if it's due to the pods not installing? When I do `pod install` I don't seen an entry for lottie at all. Is this supposed to happen?

**Comment by santinucera** [comment]
@elgambet if you downgrade the version to `2.6.1`, the android app won't build because that version doesn't support `AndroidX`. I try with every version that supports `AndroidX` and the asset  doesn't show up.

**Comment by henrykaasik** [comment]
Has anybody found a solution? How to get it working in both Android and iOS?

**Comment by JakeHadley** [comment]
Also looking for a solution. Reinstalling didn't work for me. On RN 60.4 With the recommended package versions for `lottie-react-native` and `lottie-ios`.

**Comment by WFolini** [comment]
I have the same issue. Any news about this or is there a workaround?

**Comment by greg7gkb** [comment]
Reproduced with the following minimal sandbox project:
```bash
react-native init sandbox --version react-native@0.60.6
cd sandbox
yarn add lottie-react-native
yarn add lottie-ios
cd ios
pod install
cd ..
react-native run-ios
```

Modify App.js to render:
```javascript
const App = () => {
  return (
    <LottieView style={{
      width: 300,
      height: 300,
    }} source={require('./ex.json') } autoplay loop />
  );
};
```
Running on Android shows the Lottie animation correctly, while on iOS simulator nothing appears in the LottieView.

Package dependencies resolve to:
```
    "lottie-ios": "3.1.5",
    "lottie-react-native": "3.3.2",
    "react": "16.8.6",
    "react-native": "0.60.6"
```

For this exercise, I used the animation available here: https://lottiefiles.com/11287-fashion-app-loading

**Comment by PabonSEC** [comment]
@henrykaasik , thanks a lot. It's working for me. I have used this with `react-native-animated-loader` from [here](https://www.npmjs.com/package/react-native-animated-loader)

"react": "16.9.0",
"react-native": "0.61.5",
"lottie-react-native": "^3.3.2",
"lottie-ios": "^3.1.3",
"react-native-animated-loader": "^0.0.8"

Added `use_frameworks!` into Podfile in IOS folder as well.(This actually has solved my issue).

**Comment by c-goettert** [comment]
My animations didn't play either after updating RN + Lottie to the latest versions. I am accessing the lottie-files from my asset bundle. To get it working again, I somehow needed to delete the `.json`-suffix from the asset path:
```diff
        <LottieView
-            source={'Path/to/file.json'}
+            source={'Path/to/file'}
        ...
```
Hope it helps someone.

**Comment by sushilmishraios** [comment]
constructor(props) {
        super(props);
        this.state = {
            progress: new Animated.Value(0),
        };
    }
    componentDidMount() {
        Animated.timing(this.state.progress, {
            toValue: 1,
            duration: 5000,
            easing: Easing.linear,
        }).start();
    }
    /**
     * Render View
     * */
    render() {
        return(
            <View style={{flex: 1, alignItems: 'center', justifyContent: 'center'}}>
                <LottieView
                    style={{backgroundColor: 'red'}}
                    ref={animation => {
                        this.animation = animation;
                    }}
                    source={require('../../resources/json_file/check')}
                    progress={this.state.progress}
                />
            </View>

        )

LottieView is visble but inside image with animation is not working.

**Comment by abbasmoosavi** [comment]
> Hi, also happening in:
> 
> "react-native": "0.61.2"
> "lottie-ios": "3.1.3",
> "lottie-react-native": "3.2.1",
> 
> In my case the asset I'm using works in Android and worked in IOS with previous* versions of react-native and lottie.
> 
> Other animations are working without problems.
> 
> * The animation works with:
> "react-native": "0.61.2",
> "lottie-ios": "2.5.0",
> "lottie-react-native": "2.6.1",

Great! this way worked for me! thanks my bro

**Comment by jonathanmachado** [comment]
Same here

**Comment by miketalbot** [comment]
For anyone (like me) coming across this thread later, I found the underlying problem caused by the switch to using Swift.  The JSON for the animations is being decoded in Swift and throws an error if it encounters a value it does not expect.  For me this was: "bm": 16 (blend mode 16), the library only supports blend modes 0-15 and "tp": "op" - shape type Offset Path. The Swift JSON decoder throws an error when encountering these values.  You can fix it by modifying the animation either by code or manually to remove the offending entries.  There may be others I haven't found that also trip it up.

I have written a JS function that makes the animations I've tried work (and they look fine):

```javascript
import React from 'react'
import LottieView from 'lottie-react-native'

export function clean(item) {
    for (let [key, value] of Object.entries(item)) {
        if (key === 'bm' && value > 15) {
            item.bm = 0
        } else if (Array.isArray(value)) {
            let remove = []
            for (let subItem of value) {
                if (subItem?.ty === 'op') {
                    remove.push(subItem)
                    continue
                }
                recurseClean(subItem)
            }
            for (let del of remove) {
                value.splice(value.indexOf(del), 1)
            }
        } else if (typeof value === 'object' && value) {
            recurseClean(value)
        }
    }
    return item
}

export function Lottie({ source, ...props }) {
    return <LottieView source={clean(source)} {...props} />
}
```

This also includes a component that wraps LottieView and does it on the source automatically.

If you are having problems and want to see why Lottie can't load a JSON file on IOS then you could go to the file node_modules/lottie-react-native/src/ios/LottieReactNative/ContainerView.swift and add replace @objc func setSourceJson to this - which will at least show the JSON parse error:

```swift
@objc func setSourceJson(_ newSourceJson: String) {
        sourceJson = newSourceJson
        let odata = sourceJson.data(using: String.Encoding.utf8)
        do {
            try JSONDecoder().decode(Animation.self, from: odata!)
        } catch {
            print(error)
        }
        guard let data = sourceJson.data(using: String.Encoding.utf8),
        let animation = try? JSONDecoder().decode(Animation.self, from: data) else {
            if (RCT_DEV == 1) {
                print("Unable to create the lottie animation object from the JSON source")
            }
            return
        }

        let starAnimationView = AnimationView()
        starAnimationView.animation = animation
        replaceAnimationView(next: starAnimationView)
    }
```

**Comment by Krrunal** [comment]
> > Hi, also happening in:
> > "react-native": "0.61.2"
> > "lottie-ios": "3.1.3",
> > "lottie-react-native": "3.2.1",
> > In my case the asset I'm using works in Android and worked in IOS with previous* versions of react-native and lottie.
> > Other animations are working without problems.
> > 
> > * The animation works with:
> >   "react-native": "0.61.2",
> >   "lottie-ios": "2.5.0",
> >   "lottie-react-native": "2.6.1",
> 
> Great! this way worked for me! thanks my bro

Thanks, it's works for me. Downgrading is only one solution.

**Comment by elgambet** [comment]
> For anyone (like me) coming across this thread later, I found the underlying problem caused by the switch to using Swift. The JSON for the animations is being decoded in Swift and throws an error if it encounters a value it does not expect. For me this was: "bm": 16 (blend mode 16), the library only supports blend modes 0-15 and "tp": "op" - shape type Offset Path. The Swift JSON decoder throws an error when encountering these values. You can fix it by modifying the animation either by code or manually to remove the offending entries. There may be others I haven't found that also trip it up.
> 
> I have written a JS function that makes the animations I've tried work (and they look fine):
> 
> ```js
> import React from 'react'
> import LottieView from 'lottie-react-native'
> 
> export function clean(item) {
>     for (let [key, value] of Object.entries(item)) {
>         if (key === 'bm' && value > 15) {
>             item.bm = 0
>         } else if (Array.isArray(value)) {
>             let remove = []
>             for (let subItem of value) {
>                 if (subItem?.ty === 'op') {
>                     remove.push(subItem)
>                     continue
>                 }
>                 clean(subItem)
>             }
>             for (let del of remove) {
>                 value.splice(value.indexOf(del), 1)
>             }
>         } else if (typeof value === 'object' && value) {
>             clean(value)
>         }
>     }
>     return item
> }
> 
> export function Lottie({ source, ...props }) {
>     return <LottieView source={clean(source)} {...props} />
> }
> ```
> 
> This also includes a component that wraps LottieView and does it on the source automatically.
> 
> If you are having problems and want to see why Lottie can't load a JSON file on IOS then you could go to the file node_modules/lottie-react-native/src/ios/LottieReactNative/ContainerView.swift and add replace @objc func setSourceJson to this - which will at least show the JSON parse error:
> 
> ```swift
> @objc func setSourceJson(_ newSourceJson: String) {
>         sourceJson = newSourceJson
>         let odata = sourceJson.data(using: String.Encoding.utf8)
>         do {
>             try JSONDecoder().decode(Animation.self, from: odata!)
>         } catch {
>             print(error)
>         }
>         guard let data = sourceJson.data(using: String.Encoding.utf8),
>         let animation = try? JSONDecoder().decode(Animation.self, from: data) else {
>             if (RCT_DEV == 1) {
>                 print("Unable to create the lottie animation object from the JSON source")
>             }
>             return
>         }
> 
>         let starAnimationView = AnimationView()
>         starAnimationView.animation = animation
>         replaceAnimationView(next: starAnimationView)
>     }
> ```

Hello @miketalbot  thanks for your help, using your swift code I was able to solve my problem, in my case it was that the assets array items didn't have the necessary keys "p" and "u", so taking your code as an example I created a method to clean for my asset:

```
function clean(source: {[key in string]: any}): any {
  return {
    ...source,
    assets: (source.assets as Array<any>).map(asset => ({p: '', u: '', ...asset})),
  };
}
```

**Comment by bastiantowers** [comment]
Maybe it will help others. In my case, for some reason I did not research further, the animations in my app running in the iPhone simulator was not showing at all, without throwing any errors. After strugling a bit, I deleted the app of the phone, and then reinstall the app with:
`npx react-native start --reset-cache`
`npx react-native run-ios`

**Comment by pennersr** [comment]
Thanks @miketalbot. In my case, it was caused by:

```
dataCorrupted(Swift.DecodingError.Context(codingPath: [_JSONKey(stringValue: "Index 10", intValue: 10), _JSONKey(stringValue: "Index 0", intValue: 0), _JSONKey(stringValue: "Index 0", intValue: 0), CodingKeys(stringValue: "d", intValue: nil)], debugDescription: "Cannot initialize PathDirection from invalid Int value 0", underlyingError: nil))
```

So, I opted for removing `"d"` keys with value `0`:

```
export function cleanLottie(item) {
    for (let [key, value] of Object.entries(item)) {
        if (key === 'd' && value === 0) {
          delete item.d
        } else if (key === 'bm' && value > 15) {
            item.bm = 0
        } else if (Array.isArray(value)) {
            let remove = []
            for (let subItem of value) {
                if (subItem?.ty === 'op') {
                    remove.push(subItem)
                    continue
                }
                cleanLottie(subItem)
            }
            for (let del of remove) {
                value.splice(value.indexOf(del), 1)
            }
        } else if (typeof value === 'object' && value) {
            cleanLottie(value)
        }
    }
    return item
}
```

**Comment by Jarrettgohh** [comment]
As of Expo SDK V44 (React Native V0.64.3), lottie-react-native V5.0.1 and lottie-ios V3.2.3. What I have found that works is to call the .play() method on the ref from the `LottieView`, on mount of the component. I referenced this from the official Expo Lottie documentations: https://docs.expo.dev/versions/latest/sdk/lottie/.  

const animationRef = useRef<LottieView | null>();

  useEffect(() => {
    animationRef.current?.play();
  }, []);

  return (
    <LottieView
      ref={(animation) => {
        animationRef.current = animation;
      }}
      source={require("../../assets/animations/loader1.json")}
      autoPlay
      loop
      style={{ width, height }}
    />
  );

Hope I managed to help someone!

**Comment by G2Jose** [comment]
Based on @miketalbot's suggestion above, here's a handy git patch you can apply to retrieve additional debug information, if the error is related to incompatible json being used: 


```diff
diff --git a/node_modules/lottie-react-native/src/ios/LottieReactNative/ContainerView.swift b/node_modules/lottie-react-native/src/ios/LottieReactNative/ContainerView.swift
index d05a407..fdaa1d1 100644
--- a/node_modules/lottie-react-native/src/ios/LottieReactNative/ContainerView.swift
+++ b/node_modules/lottie-react-native/src/ios/LottieReactNative/ContainerView.swift
@@ -52,7 +52,12 @@ class ContainerView: RCTView {
 
     @objc func setSourceJson(_ newSourceJson: String) {
         sourceJson = newSourceJson
-
+        let odata = sourceJson.data(using: String.Encoding.utf8)
+        do {
+            try JSONDecoder().decode(Animation.self, from: odata!)
+        } catch {
+            print(error)
+        }
         guard let data = sourceJson.data(using: String.Encoding.utf8),
         let animation = try? JSONDecoder().decode(Animation.self, from: data) else {
             if (RCT_DEBUG == 1) {

```

To use: 
- Paste the patch into a file `debug.patch`
- Run `git apply debug.patch`

In my case it turned out to be a missing `ind` key on `layers`, which I fixed with this snippet: 

```typescript
/**
 * Process bodymovin json to make it renderable by react-native-lottie
 * @param lottieJson
 */
const processJson = (lottieJson: LottieJson) => {
  return {
    ...lottieJson,
    layers: lottieJson.layers.map(layer => ({
      ind: 0,
      ...layer,
    })),
  } as FIXME
}
```

**Comment by a-j-douglass** [comment]
@G2Jose this might be a stupid question but if I apply your patch (or add any `prints` in ContainerView.swift ) what do I have to do to see the logging?

I've tried refreshing and making a new eas build, and looked in expo's output and in the system logs but I don't get anything.

**Comment by G2Jose** [comment]
@a-j-douglass it sounds like you're using expo which I'm not as familiar with. See if [this section](https://docs.expo.dev/workflow/logging/#native-logs) helps at all. 

These logs typically show on xcode on a bare (non expo) react native build.

**Comment by timoisalive** [comment]
I'm having issues playing back animations on my iOS production build, whereas on Expo Go (in iOS Simulator or device) the animation plays fine. The JSON seems to be there on production build as well, as I'm logging the length of the JSON string, but nothing shows, not even paused animation... I tried @Jarrettgohh 's suggestion, but that didn't help either. 🤷 

I've successfully used this library in another app, but that was probably a few versions back and a different setup, so not easily comparable.
